packages/angular/cli/tasks/install-package.ts

```diff
@@ -0,0 +1,182 @@
+ */
+
+import { logging } from '@angular-devkit/core';
+import { spawnSync } from 'child_process';
+import {
+  existsSync,
+  mkdtempSync,
+  readFileSync,
+  realpathSync,
+} from 'fs';
+import { tmpdir } from 'os';
+import { join, resolve } from 'path';
+import * as rimraf from 'rimraf';
+import { PackageManager } from '../lib/config/schema';
+import { colors } from '../utilities/color';
+import { NgAddSaveDepedency } from '../utilities/package-metadata';
+
+interface PackageManagerOptions {
+  silent: string;
+  saveDev: string;
+  install: string;
+  prefix: string;
+}
+
+export function installPackage(
+  packageName: string,
+  logger: logging.Logger,
+  packageManager: PackageManager = PackageManager.Npm,
+  save: Exclude<NgAddSaveDepedency, false> = true,
+  extraArgs: string[] = [],
+  global = false,
+) {
+  const packageManagerArgs = getPackageManagerArguments(packageManager);
+
+  const installArgs: string[] = [
+    packageManagerArgs.install,
+    packageName,
+    packageManagerArgs.silent,
+  ];
+
+  logger.info(colors.green(`Installing packages for tooling via ${packageManager}.`));
+
+  if (save === 'devDependencies') {
+    installArgs.push(packageManagerArgs.saveDev);
+  }
+
+  if (global) {
+    if (packageManager === PackageManager.Yarn) {
+      installArgs.unshift('global');
+    } else {
+      installArgs.push('--global');
+    }
+  }
+
+  const { status } = spawnSync(
+    packageManager,
+    [
+      ...installArgs,
+      ...extraArgs,
+    ],
+    {
+      stdio: 'inherit',
+      shell: true,
+    },
+  );
+
+  if (status !== 0) {
+    throw new Error('Package install failed, see above.');
+  }
+
+  logger.info(colors.green(`Installed packages for tooling via ${packageManager}.`));
+}
+
+export function installTempPackage(
+  packageName: string,
+  logger: logging.Logger,
+  packageManager: PackageManager = PackageManager.Npm,
+): string {
+  const tempPath = mkdtempSync(join(realpathSync(tmpdir()), '.ng-temp-packages-'));
+
+  // clean up temp directory on process exit
+  process.on('exit', () => rimraf.sync(tempPath));
+
+  // setup prefix/global modules path
+  const packageManagerArgs = getPackageManagerArguments(packageManager);
+  const installArgs: string[] = [
+    packageManagerArgs.prefix,
+    tempPath,
+  ];
+
+  installPackage(packageName, logger, packageManager, true, installArgs, true);
+
+  let tempNodeModules: string;
+  if (packageManager !== PackageManager.Yarn && process.platform !== 'win32') {
+    // Global installs on Unix systems go to {prefix}/lib/node_modules.
+    // Global installs on Windows go to {prefix}/node_modules (that is, no lib folder.)
+    tempNodeModules = join(tempPath, 'lib', 'node_modules');
+  } else {
+    tempNodeModules = join(tempPath, 'node_modules');
+  }
+
+  // Needed to resolve schematics from this location since we use a custom
+  // resolve strategy in '@angular/devkit-core/node'
+  process.env.NG_TEMP_MODULES_DIR = tempNodeModules;
```

Comment:
The schematics resolution is the last place the custom resolver is being used.  We should move it to native methods when possible.  Can you add a note that this won't be needed when that refactor is done?

Reply:
will do, I actually have a following PR to remove that logic. Will try to land it in the coming days.